### PR TITLE
Tree based trainers implement ICanGetSummaryAsIDataView

### LINKED ISSUE #3755: Model summary to show tree details for FastTree
Currently model summary doesnt show trees and split points for FastTree models.
FastTree model needs to implement ICanGetSummaryAsIDataView that would dump this info.

### PR DESCRIPTION
**Pull request by artidoro**
Fixes #3755.

NimbusML did not have access to the details on the tree structure.
This PR implements the `ICanGetSummaryAsIDataView` interface which is used in the `Summarize` entrypoint to pass a summary of the model parameters to NimbusML in the form of an `IDataView`.

I create a utility method that does the conversion from `RegressionTreeBase` to `IDataView` with special treatment for `QuantileRegressionTrees` which have additional information.

In the `IDataView`, each node has its own row and the columns correspond to the fields describing each node. To determine which tree the node belongs to there is a `TreeID` column.

### CODE REVIEW

File: test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs [hunk 1]
```diff
@@ -970,8 +970,8 @@ public void EntryPointPipelineEnsembleText()
                 var data = splitOutput.TrainData[i];
                 if (i % 2 == 0)
                 {
-                    data = new TextFeaturizingEstimator(Env, "Features", new List<string> { "Text" }, 
-                        new TextFeaturizingEstimator.Options { 
+                    data = new TextFeaturizingEstimator(Env, "Features", new List<string> { "Text" },
```
Comment (artidoro) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs]: Revert changes not related to PR.

File: src/Microsoft.ML.FastTree/FastTree.cs [hunk 2]
```diff
@@ -3378,7 +3361,7 @@ public TreeNode(Dictionary<string, object> keyValues)
     /// and <see cref="TreeEnsembleModelParametersBasedOnRegressionTree"/> is the type of
     /// <see cref="TrainedTreeEnsemble"/>.
     /// </summary>
-    public abstract class TreeEnsembleModelParametersBasedOnRegressionTree : TreeEnsembleModelParameters
+    public abstract class TreeEnsembleModelParametersBasedOnRegressionTree : TreeEnsembleModelParameters, ICanGetSummaryAsIDataView
```
Comment (wschin) [inline src/Microsoft.ML.FastTree/FastTree.cs]: Why do we need `ICanGetSummaryAs...`? `IDataView RegressionTreeEnsembleAsIDataView(..)` looks sufficient for generating a summary.

**Comment by ganik** [comment]
>         public void EntryPointPipelineEnsembleText()

Pls add test using graph json text

---
Refers to: test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs:934 in 81a5d39. [](commit_id = 81a5d39dccd82c8bc440fd139202e458e2adee1e, deletion_comment = False)

File: src/Microsoft.ML.FastTree/FastTree.cs (hunk 2 above)
Comment (artidoro) [reply src/Microsoft.ML.FastTree/FastTree.cs]: That's the interface that Summarize entrypoint uses to get the IDataView from a trainer.

See: https://github.com/dotnet/machinelearning/blob/665a3669652bc478fffbc7ed73b89067e5a7938c/src/Microsoft.ML.Data/EntryPoints/SummarizePredictor.cs#L32

And:
https://github.com/dotnet/machinelearning/blob/665a3669652bc478fffbc7ed73b89067e5a7938c/src/Microsoft.ML.Data/EntryPoints/SummarizePredictor.cs#L49

File: src/Microsoft.ML.FastTree/FastTree.cs [hunk 3]
```diff
@@ -3406,6 +3389,12 @@ private RegressionTreeEnsemble CreateTreeEnsembleFromInternalDataStructure()
             var treeWeights = TrainedEnsemble.Trees.Select(tree => tree.Weight);
             return new RegressionTreeEnsemble(trees, treeWeights, TrainedEnsemble.Bias);
         }
+
+        /// <summary>
+        /// Used for the Summarize entrypoint.
+        /// </summary>
+        IDataView ICanGetSummaryAsIDataView.GetSummaryDataView(RoleMappedSchema schema)
+        => RegressionTreeBaseUtils.RegressionTreeEnsembleAsIDataView(Host, TrainedTreeEnsemble.Bias, TrainedTreeEnsemble.TreeWeights, TrainedTreeEnsemble.Trees);
```
Comment (artidoro) [inline src/Microsoft.ML.FastTree/FastTree.cs]: indent

File: test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs [hunk 4]
```diff
@@ -597,5 +598,247 @@ public void LightGbmInDifferentCulture()
...
+            Assert.NotNull(modelParameters);
+
+            CheckSummary(modelParameters, trainedTreeEnsemble.Bias, trainedTreeEnsemble.TreeWeights, trainedTreeEnsemble.Trees);
+            Done();
+        }
+
+        [LightGBMFact]
+        public void LightGbmBinaryClassificationTestSummary()
+        {
+            var (pipeline, dataView) = GetBinaryClassificationPipeline();
```
Comment (artidoro) [inline test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs]: >GetBinaryClassificationPipeline [](start = 39, length = 31)

Make sure that these work with the categorical split after one hot encoding.

**Pull request merged**